crates/router/src/core/payment_methods/cards.rs

```diff
@@ -1691,12 +1691,39 @@ pub async fn list_payment_methods(
     let db = &*state.store;
     let pm_config_mapping = &state.conf.pm_filters;
 
-    let payment_intent = helpers::verify_payment_intent_time_and_client_secret(
-        db,
-        &merchant_account,
-        req.client_secret.clone(),
-    )
-    .await?;
+    let payment_intent = if let Some(cs) = &req.client_secret {
+        if cs.starts_with("pm_") {
+            let pm_vec = cs.split("_secret").collect::<Vec<&str>>();
+            let pm_id = pm_vec
+                .first()
+                .ok_or(errors::ApiErrorResponse::MissingRequiredField {
+                    field_name: "client_secret",
+                })?;
+
+            let payment_method = db
+                .find_payment_method(pm_id, merchant_account.storage_scheme)
+                .await
+                .change_context(errors::ApiErrorResponse::PaymentMethodNotFound)
+                .attach_printable("Unable to find payment method")?;
+
+            let client_secret_expired =
+                authenticate_pm_client_secret_and_check_expiry(cs, &payment_method)?;
+            if client_secret_expired {
+                return Err((errors::ApiErrorResponse::ClientSecretExpired).into());
+            }
+
```

Comment:
Are we sure it is okay to not return early here?

Reply:
Could you elaborate a bit on this?